numpy/core/src/multiarray/ctors.c

```diff
@@ -1344,18 +1390,41 @@ _array_from_buffer_3118(PyObject *obj, PyObject **out)
         }
     }
 
-    flags = NPY_ARRAY_BEHAVED & (view->readonly ? ~NPY_ARRAY_WRITEABLE : ~0);
-    r = PyArray_NewFromDescr(&PyArray_Type, descr,
-                             nd, shape, strides, view->buf,
-                             flags, NULL);
-    if (r == NULL ||
-            PyArray_SetBaseObject((PyArrayObject *)r, memoryview) < 0) {
-        Py_XDECREF(r);
-        Py_DECREF(memoryview);
-        return -1;
+    if (view->suboffsets == NULL) {
+         flags = NPY_ARRAY_BEHAVED & (view->readonly ? ~NPY_ARRAY_WRITEABLE : ~0);
+         r = PyArray_NewFromDescr(&PyArray_Type, descr,
+                                  nd, shape, strides, view->buf,
+                                  flags, NULL);
+         if (r == NULL ||
+             PyArray_SetBaseObject((PyArrayObject *)r, memoryview) < 0) {
+              Py_XDECREF(r);
+              Py_DECREF(memoryview);
+              return -1;
+         }
+         PyArray_UpdateFlags((PyArrayObject *)r, NPY_ARRAY_UPDATE_ALL);
     }
-    PyArray_UpdateFlags((PyArrayObject *)r, NPY_ARRAY_UPDATE_ALL);
+    else {
+         /* Recompute strides */
+         npy_intp blocksize = view->itemsize;
+         for (k = nd - 1; k >= 0; --k) {
+             if (view->suboffsets[k] >= 0) {
+                strides[k] = blocksize;
+             }
+             blocksize *= shape[k];
+         }
+         r = PyArray_NewFromDescr(&PyArray_Type, descr,
+                                  nd, shape, strides, NULL, 0, NULL);
+         if (r == NULL)
+            goto fail;
+         if (PyArray_SetBaseObject((PyArrayObject *)r, memoryview) < 0) {
+            Py_DECREF(r);
+            goto fail;
```

Comment:
[PyArray_NewFromDescr](http://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_NewFromDescr) steals a reference to `descr`, so the goto's above should not got to 

```
 fail:
     Py_XDECREF(descr);		     Py_XDECREF(descr);
```